### PR TITLE
Add iswitness parameter to decode- and fundrawtransaction RPCs

### PR DESCRIPTION
**Pull request by meshcollider**
Suggested in https://github.com/bitcoin/bitcoin/pull/10481#issuecomment-325244946, this adds the option to explicitly choose whether a serialized transaction should be decoded as a witness or non-witness transaction rather than relying on the heuristic checks in #10481. The parameter defaults to relying on #10481 if not included, but it overrides that if included.

### CODE REVIEW

**Comment by MeshCollider** [comment]
@achow101 @roconnor-blockstream you might want to take a look at this :)

File: src/core_read.cpp [hunk 1]
```diff
@@ -108,7 +108,7 @@ bool CheckTxScriptsSanity(const CMutableTransaction& tx)
     return true;
 }
 
-bool DecodeHexTx(CMutableTransaction& tx, const std::string& strHexTx, bool fTryNoWitness)
+bool DecodeHexTx(CMutableTransaction& tx, const std::string& strHexTx, bool fTryNoWitness, bool fTryWitness)
```
Comment (achow101) [inline src/core_read.cpp]: `snake_case`

**Review by achow101 — COMMENTED** [review]
utACK 55e2b9c8a13b5f80d03936f4275cd9e3828e4b3f

**Comment by MeshCollider** [comment]
Changed style as requested, thanks @achow101

**Comment by luke-jr** [comment]
Please don't change the style of *existing* code, only the new code... :/

**Comment by MeshCollider** [comment]
@luke-jr the style refactor only touches a tiny amount of code inside that single function, in a similar way to adding braces to existing if-statements when modifying nearby code. I'd argue that this small change to existing code style is worth it for the consistency, especially between try_witness and try_no_witness

**Review by promag — COMMENTED** [review]
Missing tests.

File: src/core_read.cpp [hunk 2]
```diff
@@ -108,38 +108,44 @@ bool CheckTxScriptsSanity(const CMutableTransaction& tx)
...
+        catch (const std::exception&) {
             return false;
         }
     }
-    catch (const std::exception&) {
-        return false;
-    }
-
+    
     return true;
```
Comment (sipa) [inline src/core_read.cpp]: If you move this `return true;` inside the `if (try_witness)` block, the `if(!try_witness)` inside the `catch` above becomes unnecessary. I think that'd be a cleaner flow.

File: src/rpc/rawtransaction.cpp [hunk 3]
```diff
@@ -418,13 +418,15 @@ UniValue createrawtransaction(const JSONRPCRequest& request)
...
-    if (request.fHelp || request.params.size() != 1)
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
         throw std::runtime_error(
-            "decoderawtransaction \"hexstring\"\n"
+            "decoderawtransaction \"hexstring\" ( iswitness )\n"
             "\nReturn a JSON object representing the serialized, hex-encoded transaction.\n"
 
             "\nArguments:\n"
             "1. \"hexstring\"      (string, required) The transaction hex string\n"
+            "2. iswitness          (boolean, optional) Whether the transaction hex is a serialized witness transaction \n"
```
Comment (sipa) [inline src/rpc/rawtransaction.cpp]: Nit: trailing space

Comment (MeshCollider) [reply src/rpc/rawtransaction.cpp]: Fixed

File: src/core_read.cpp (hunk 2 above)
Comment (MeshCollider) [reply src/core_read.cpp]: Good idea yep, done

File: src/core_read.cpp [hunk 4]
```diff
@@ -129,18 +129,22 @@ bool DecodeHexTx(CMutableTransaction& tx, const std::string& strHexTx, bool fTry
...
-        if (!ssData.empty()) {
+    if (try_witness) {
+        CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
+        try {
+            ssData >> tx;
+            if (!ssData.empty()) {
+                return false;
+            }
+        }
+        catch (const std::exception&) {
```
Comment (sipa) [inline src/core_read.cpp]: Nit: `catch` on the same line as `}`.

File: src/core_read.cpp [hunk 5]
```diff
@@ -129,18 +129,22 @@ bool DecodeHexTx(CMutableTransaction& tx, const std::string& strHexTx, bool fTry
...
-    CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
-    try {
-        ssData >> tx;
-        if (!ssData.empty()) {
+    if (try_witness) {
+        CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
+        try {
+            ssData >> tx;
+            if (!ssData.empty()) {
+                return false;
```
Comment (sipa) [inline src/core_read.cpp]: Even shorter: invert this test and make it return true.

Comment (MeshCollider) [reply src/core_read.cpp]: Nice :+1:

File: src/core_read.cpp [hunk 6]
```diff
@@ -108,39 +108,39 @@ bool CheckTxScriptsSanity(const CMutableTransaction& tx)
...
-        if (!ssData.empty()) {
+    if (try_witness) {
+        CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION);
+        try {
+            ssData >> tx;
+            if (ssData.empty()) {
+                return true;
+            }
+        } catch (const std::exception&) {
             return false;
```
Comment (sipa) [inline src/core_read.cpp]: And now this line is unnecessary as well :)

Comment (MeshCollider) [reply src/core_read.cpp]: Heh, changed. Next you'll want me to move the try-catch outside the if statements so there only needs to be one of them ;) Thanks for reviewing!

Comment (sipa) [reply src/core_read.cpp]: Nope, that wouldn't actually work. If both flags were true, and the first attempt to deserialize throws, you still want to try the second.

File: src/core_read.cpp [hunk 7]
```diff
@@ -108,39 +108,39 @@ bool CheckTxScriptsSanity(const CMutableTransaction& tx)
...
-    std::vector<unsigned char> txData(ParseHex(strHexTx));
+    std::vector<unsigned char> txData(ParseHex(hex_tx));
 
-    if (fTryNoWitness) {
+    if (try_no_witness) {
         CDataStream ssData(txData, SER_NETWORK, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS);
         try {
             ssData >> tx;
-            if (ssData.eof() && CheckTxScriptsSanity(tx)) {
+            if (ssData.eof() && (CheckTxScriptsSanity(tx) || !try_witness)) {
```
Comment (sipa) [inline src/core_read.cpp]: One final tiny nit: swap the arguments, as checking a boolean is much faster than CheckTxScriptsSanity.

**Comment by sipa** [comment]
utACK 3d16658226cd08911200dbd90db762dc3409e443

**Comment by MeshCollider** [comment]
Rebased

**Review by ryanofsky — COMMENTED** [review]
utACK 6f39ac04375a5f6ef803da59ba0b606123d63142

Probably could merge given 3 acks.

**Comment by laanwj** [comment]
utACK 6f39ac04375a5f6ef803da59ba0b606123d63142

**Pull request merged**